File: camera_stream.py
#!/usr/bin/env python3
"""
Módulo para manejo de streaming de cámaras RTSP → HLS con ffmpeg.
Provee utilidades que usa el servidor:
- asegurar_carpetas()
- construir_rtsp_url(config, indice, calidad)
- obtener_resolucion(config, cam_id)
- obtener_bitrate(resolucion)
- iniciar_hls(cam_id, url_rtsp)
- detener_hls(cam_id)
- detener_todos()
"""

import logging
import os
import time
from pathlib import Path
import shutil
import signal
import subprocess
from typing import Dict, Optional, Tuple

logger = logging.getLogger(__name__)

# Configuración
HLS_OUTPUT_DIR = Path(__file__).parent / "hls"
HLS_PLAYLIST_NAME = {
    "cam1": "cam1.m3u8",
    "cam2": "cam2.m3u8",
}

# Mapa de resoluciones a escalas y bitrates
RESOLUCION_CONFIG = {
    "360p": {"scale": "640:360", "bitrate": "1000k", "maxrate": "1500k"},
    "480p": {"scale": "854:480", "bitrate": "1500k", "maxrate": "2000k"},
    "720p": {"scale": "1280:720", "bitrate": "2500k", "maxrate": "3500k"},
    "1080p": {"scale": "1920:1080", "bitrate": "4000k", "maxrate": "5000k"},
}

# Registro de procesos ffmpeg por cámara
PROCESOS: Dict[str, subprocess.Popen] = {}

def asegurar_carpetas():
    """Crea las carpetas necesarias para HLS."""
    try:
        HLS_OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
        for cam in ("cam1", "cam2"):
            (HLS_OUTPUT_DIR / cam).mkdir(parents=True, exist_ok=True)
        logger.info("✓ Carpetas HLS listas")
        return True
    except Exception as e:
        logger.error(f"✗ Error creando carpetas HLS: {e}")
        return False


def obtener_resolucion(config: dict, cam_id: str, velocidad_kbps: float = None) -> str:
    """Obtiene la resolución configurada para una cámara.
    
    Args:
        config: Diccionario de configuración
        cam_id: 'cam1' o 'cam2'
        velocidad_kbps: Velocidad de red en Kbps (opcional, para modo automático)
    
    Returns:
        str: Resolución (ej: '480p', '720p')
    """
    indice = 1 if cam_id == 'cam1' else 2
    modo = config.get(f'camara{indice}_modo_resolucion', 'manual')
    
    if modo == 'automatico':
        # Modo automático: ajustar según velocidad de red
        if velocidad_kbps is None:
            # Si no tenemos velocidad, usar resolución manual o default
            return config.get(f'camara{indice}_resolucion', '480p')
        
        # Lógica automática basada en velocidad de internet
        # Convertir Kbps a Mbps para cálculo más claro
        velocidad_mbps = velocidad_kbps / 1000.0
        
        if velocidad_mbps < 2.0:
            resolucion = '360p'  # Conexión lenta
        elif velocidad_mbps < 5.0:
            resolucion = '480p'  # Conexión media
        elif velocidad_mbps < 10.0:
            resolucion = '720p'  # Conexión buena
        else:
            resolucion = '1080p'  # Conexión excelente
        
        logger.info(f"Modo automático {cam_id}: {velocidad_mbps:.1f} Mbps → {resolucion}")
        return resolucion
    else:
        # Modo manual
        resolucion = config.get(f'camara{indice}_resolucion', '480p')
        if resolucion not in RESOLUCION_CONFIG:
            resolucion = '480p'
        return resolucion


def obtener_bitrate(resolucion: str) -> Tuple[str, str]:
    """Obtiene bitrate y maxrate para una resolución.
    
    Returns:
        tuple: (bitrate, maxrate)
    """
    if resolucion not in RESOLUCION_CONFIG:
        resolucion = '480p'
    config = RESOLUCION_CONFIG[resolucion]
    return config['bitrate'], config['maxrate']


def construir_rtsp_url(config: dict, indice: int, calidad: str = "sd") -> Optional[str]:
    """Devuelve la primera URL RTSP candidata para compatibilidad hacia atrás.

    Mantiene la firma antigua, pero ahora delega a `construir_rtsp_candidatos`
    y devuelve únicamente la primera URL disponible.
    """
    candidatos = construir_rtsp_candidatos(config, indice)
    return candidatos[0] if candidatos else None


def construir_rtsp_candidatos(config: dict, indice: int) -> list[str]:
    """Construye una lista de URLs RTSP a partir de los parámetros ONVIF o URL completa.

    - Si existe rtsp_camaraN_url, la usa directamente como primera candidata.
    - Si no, construye URLs desde host/puerto/usuario/contraseña/perfil ONVIF.
    - Genera candidatos probando puertos habituales: 554 (RTSP) y el puerto
      ONVIF indicado (por defecto 8899) si es diferente.
    """
    try:
        if not config:
            return []

        # Obtener URL RTSP completa si existe (prioritaria)
        url_completa = ""
        if indice == 1:
            url_completa = (config.get("rtsp_camara1_url") or "").strip()
        else:
            url_completa = (config.get("rtsp_camara2_url") or "").strip()
        
        # Si hay URL completa, usarla como primera opción
        candidatos = []
        if url_completa and url_completa.startswith("rtsp://"):
            candidatos.append(url_completa)

        # Construir URLs desde parámetros ONVIF individuales
        if indice == 1:
            host = (config.get("onvif_camara1_host") or "").strip()
            puerto_onvif = int(config.get("onvif_camara1_puerto", 8899) or 8899)
            puerto_rtsp = int(config.get("rtsp_camara1_puerto", 554) or 554)
            usuario = (config.get("onvif_camara1_usuario") or "").strip()
            contrasena = (config.get("onvif_camara1_contrasena") or "").strip()
            perfil = (config.get("onvif_camara1_perfil") or "Streaming/Channels/101").strip("/")
            logger.debug(f"CAM1 DEBUG: host={host}, usuario={usuario}, contrasena={contrasena}, perfil={perfil}")
        else:
            host = (config.get("onvif_camara2_host") or "").strip()
            puerto_onvif = int(config.get("onvif_camara2_puerto", 8899) or 8899)
            puerto_rtsp = int(config.get("rtsp_camara2_puerto", 554) or 554)
            usuario = (config.get("onvif_camara2_usuario") or "").strip()
            contrasena = (config.get("onvif_camara2_contrasena") or "").strip()
            perfil = (config.get("onvif_camara2_perfil") or "Streaming/Channels/101").strip("/")

        if host:
            auth = ""
            if usuario:
                auth = usuario
                if contrasena:
                    auth += f":{contrasena}"
                auth += "@"

            puertos = [puerto_rtsp]  # Prioridad: puerto RTSP especificado
            if puerto_onvif not in puertos:
                puertos.append(puerto_onvif)
            if 554 not in puertos:
                puertos.append(554)  # Puerto RTSP estándar

            # Agregar candidatos desde parámetros ONVIF (si no están ya)
            for p in puertos:
                url = f"rtsp://{auth}{host}:{p}/{perfil}"
                if url not in candidatos:
                    candidatos.append(url)

        return candidatos
    except Exception as e:
        logger.error(f"✗ Error construyendo candidatos RTSP desde ONVIF: {e}")
        return []

def _resolver_cam_id(indice_o_id) -> Optional[str]:
    if isinstance(indice_o_id, str):
        if indice_o_id in ("cam1", "cam2"):
            return indice_o_id
        return None
    try:
        i = int(indice_o_id)
        return "cam1" if i == 1 else ("cam2" if i == 2 else None)
    except Exception:
        return None


def _limpiar_salida(cam_id: str) -> None:
    carpeta = HLS_OUTPUT_DIR / cam_id
    try:
        for p in carpeta.glob("*.ts"):
            p.unlink(missing_ok=True)
        for p in carpeta.glob("*.m3u8"):
            p.unlink(missing_ok=True)
    except Exception:
        pass


def _iniciar_hls_single(cam_id: str, url_rtsp: str, config: dict = None, velocidad_kbps: float = None) -> Tuple[bool, str]:
    """Lanza ffmpeg para una URL específica.
    
    Args:
        cam_id: 'cam1' o 'cam2'
        url_rtsp: URL RTSP de la cámara
        config: Diccionario de configuración (para obtener resolución)
        velocidad_kbps: Velocidad de red en Kbps (para modo automático)
    """
    # Verificar ffmpeg
    ffmpeg = shutil.which("ffmpeg")
    if not ffmpeg:
        logger.error("✗ ffmpeg no encontrado en el sistema. Instálelo para HLS.")
        return False, "ffmpeg no encontrado en el sistema"

    asegurar_carpetas()

    # Detener si ya hay uno
    if cam_id in PROCESOS:
        detener_hls(cam_id)

    carpeta = HLS_OUTPUT_DIR / cam_id
    _limpiar_salida(cam_id)

    playlist_path = carpeta / HLS_PLAYLIST_NAME.get(cam_id, f"{cam_id}.m3u8")
    segment_pattern = carpeta / f"{cam_id}_%03d.ts"

    # Obtener resolución configurada
    resolucion = obtener_resolucion(config or {}, cam_id, velocidad_kbps)
    bitrate, maxrate = obtener_bitrate(resolucion)
    escala = RESOLUCION_CONFIG[resolucion]['scale']
    
    logger.info(f"HLS {cam_id}: usando {resolucion} ({escala}, {bitrate})")

    cmd = [
        ffmpeg,
        "-nostdin",
        "-loglevel", "error",
        "-rtsp_transport", "tcp",
        "-i", url_rtsp,
        "-an",
        "-vf", f"scale={escala}",
        "-c:v", "libx264",
        "-preset", "ultrafast",
        "-b:v", bitrate,
        "-maxrate", maxrate,
        "-bufsize", "1000k",
        "-f", "hls",
        "-hls_time", "0.5",
        "-hls_list_size", "3",
        "-hls_flags", "delete_segments+program_date_time",
        "-hls_segment_filename", str(segment_pattern),
        str(playlist_path),
    ]

    proc = subprocess.Popen(cmd)
    PROCESOS[cam_id] = proc
    logger.info(f"✓ Iniciando HLS {cam_id} desde {url_rtsp} (PID {proc.pid})")

    for _ in range(6):
        time.sleep(0.5)
        if proc.poll() is not None:
            PROCESOS.pop(cam_id, None)
            logger.error(f"✗ ffmpeg terminó temprano para {cam_id} (código {proc.returncode})")
            return False, "No se pudo abrir el RTSP (proceso terminó)"
        if playlist_path.exists() and playlist_path.stat().st_size > 0:
            return True, "HLS iniciado"
    return True, "HLS iniciado (validación pendiente)"


def iniciar_hls(indice_o_id, url_rtsp: Optional[str | list[str]], config: dict = None, velocidad_kbps: float = None) -> Tuple[bool, str]:
    """Inicia un proceso ffmpeg que convierte RTSP a HLS.

    Permite una lista de URLs candidatas; intentará en orden hasta que una funcione.
    
    Args:
        indice_o_id: 1, 2, 'cam1' o 'cam2'
        url_rtsp: URL RTSP o lista de candidatos
        config: Diccionario de configuración (para resolución)
        velocidad_kbps: Velocidad de red en Kbps (para modo automático)
    """
    cam_id = _resolver_cam_id(indice_o_id)
    if not cam_id:
        logger.error(f"✗ Identificador de cámara inválido: {indice_o_id}")
        return False, "Identificador de cámara inválido"

    candidatos = []
    if isinstance(url_rtsp, (list, tuple)):
        candidatos = [u for u in url_rtsp if u]
    elif url_rtsp:
        candidatos = [url_rtsp]

    if not candidatos:
        logger.warning(f"⚠ Sin URLs RTSP candidatas para {cam_id}")
        return False, "URL RTSP vacía"

    ultimo_error = "URL RTSP no válida"
    for idx, url in enumerate(candidatos, start=1):
        try:
            ok, msg = _iniciar_hls_single(cam_id, url, config, velocidad_kbps)
            if ok:
                if idx > 1:
                    msg = f"{msg} (usando candidato {idx})"
                return True, msg
            ultimo_error = msg
        except Exception as e:  # pragma: no cover (log de robustez)
            ultimo_error = str(e)
            logger.error(f"✗ Error iniciando HLS para {cam_id} con candidato {idx}: {e}")
    return False, ultimo_error

def detener_hls(indice_o_id) -> bool:
    """Detiene el ffmpeg de una cámara y libera recursos."""
    cam_id = _resolver_cam_id(indice_o_id)
    if not cam_id:
        return False
    try:
        proc = PROCESOS.get(cam_id)
        if not proc:
            return True
        try:
            proc.send_signal(signal.SIGTERM)
            proc.wait(timeout=5)
        except Exception:
            try:
                proc.kill()
            except Exception:
                pass
        finally:
            PROCESOS.pop(cam_id, None)
        logger.info(f"✓ HLS detenido para {cam_id}")
        return True
    except Exception as e:
        logger.error(f"✗ Error deteniendo HLS para {cam_id}: {e}")
        return False

def detener_todos() -> bool:
    """Detiene todos los procesos ffmpeg activos."""
    ok = True
    for cam_id in list(PROCESOS.keys()):
        ok = detener_hls(cam_id) and ok
    return ok

if __name__ == "__main__":
    asegurar_carpetas()
    print("✓ camera_stream.py cargado correctamente")

File: camera_stream.go
package main

// ==================== STREAMING CÁMARAS RTSP → HLS ====================
// Gestiona procesos ffmpeg para convertir RTSP a HLS
// Sin librerías externas

import (
	"fmt"
	"log"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"
)

// ==================== CONFIGURACIÓN ====================

var hlsOutputDir string

var hlsPlaylistName = map[string]string{
	"cam1": "cam1.m3u8",
	"cam2": "cam2.m3u8",
}

// Resoluciones disponibles
var resolucionConfig = map[string]struct {
	Scale   string
	Bitrate string
	Maxrate string
}{
	"360p":  {"640:360", "1000k", "1500k"},
	"480p":  {"854:480", "1500k", "2000k"},
	"720p":  {"1280:720", "2500k", "3500k"},
	"1080p": {"1920:1080", "4000k", "5000k"},
}

// Procesos ffmpeg activos
var (
	procesosHLS   = map[string]*exec.Cmd{}
	procesosHLSMu sync.Mutex
)

func init() {
	dir, _ := os.Executable()
	hlsOutputDir = filepath.Join(filepath.Dir(dir), "hls")
	if _, err := os.Stat(hlsOutputDir); os.IsNotExist(err) {
		hlsOutputDir = "hls"
	}
}

// ==================== FUNCIONES PRINCIPALES ====================

// hlsAsegurarCarpetas crea las carpetas necesarias para HLS
func hlsAsegurarCarpetas() bool {
	if err := os.MkdirAll(hlsOutputDir, 0755); err != nil {
		log.Printf("Error creando carpeta HLS: %v", err)
		return false
	}
	for _, cam := range []string{"cam1", "cam2"} {
		if err := os.MkdirAll(filepath.Join(hlsOutputDir, cam), 0755); err != nil {
			log.Printf("Error creando carpeta HLS %s: %v", cam, err)
			return false
		}
	}
	log.Println("Carpetas HLS listas")
	return true
}

// obtenerResolucion obtiene la resolución configurada para una cámara
func obtenerResolucion(config map[string]interface{}, camID string, velocidadKbps *float64) string {
	indice := 1
	if camID == "cam2" {
		indice = 2
	}

	modoKey := fmt.Sprintf("camara%d_modo_resolucion", indice)
	resKey := fmt.Sprintf("camara%d_resolucion", indice)

	modo, _ := config[modoKey].(string)

	if modo == "automatico" {
		if velocidadKbps == nil {
			res, _ := config[resKey].(string)
			if res == "" {
				res = "480p"
			}
			return res
		}

		velocidadMbps := *velocidadKbps / 1000.0
		switch {
		case velocidadMbps < 2.0:
			return "360p"
		case velocidadMbps < 5.0:
			return "480p"
		case velocidadMbps < 10.0:
			return "720p"
		default:
			return "1080p"
		}
	}

	// Modo manual
	res, _ := config[resKey].(string)
	if res == "" {
		res = "480p"
	}
	if _, ok := resolucionConfig[res]; !ok {
		res = "480p"
	}
	return res
}

// construirRTSPCandidatos construye una lista de URLs RTSP candidatas
func construirRTSPCandidatos(config map[string]interface{}, indice int) []string {
	if config == nil {
		return nil
	}

	var candidatos []string

	// URL RTSP completa (prioritaria)
	urlKey := fmt.Sprintf("rtsp_camara%d_url", indice)
	if urlCompleta, ok := config[urlKey].(string); ok && strings.HasPrefix(strings.TrimSpace(urlCompleta), "rtsp://") {
		candidatos = append(candidatos, strings.TrimSpace(urlCompleta))
	}

	// Construir desde parámetros ONVIF
	hostKey := fmt.Sprintf("onvif_camara%d_host", indice)
	host, _ := config[hostKey].(string)
	host = strings.TrimSpace(host)

	if host != "" {
		puertoONVIFKey := fmt.Sprintf("onvif_camara%d_puerto", indice)
		puertoRTSPKey := fmt.Sprintf("rtsp_camara%d_puerto", indice)
		usuarioKey := fmt.Sprintf("onvif_camara%d_usuario", indice)
		contrasenaKey := fmt.Sprintf("onvif_camara%d_contrasena", indice)
		perfilKey := fmt.Sprintf("onvif_camara%d_perfil", indice)

		puertoONVIF := toInt(config[puertoONVIFKey])
		if puertoONVIF == 0 {
			puertoONVIF = 8899
		}
		puertoRTSP := toInt(config[puertoRTSPKey])
		if puertoRTSP == 0 {
			puertoRTSP = 554
		}

		usuario, _ := config[usuarioKey].(string)
		contrasena, _ := config[contrasenaKey].(string)
		perfil, _ := config[perfilKey].(string)
		if perfil == "" {
			perfil = "Streaming/Channels/101"
		}
		perfil = strings.Trim(perfil, "/")

		auth := ""
		if usuario != "" {
			auth = usuario
			if contrasena != "" {
				auth += ":" + contrasena
			}
			auth += "@"
		}

		// Generar candidatos con diferentes puertos
		puertos := []int{puertoRTSP}
		if puertoONVIF != puertoRTSP {
			puertos = append(puertos, puertoONVIF)
		}
		if puertoRTSP != 554 && puertoONVIF != 554 {
			puertos = append(puertos, 554)
		}

		for _, p := range puertos {
			url := fmt.Sprintf("rtsp://%s%s:%d/%s", auth, host, p, perfil)
			// Evitar duplicados
			encontrado := false
			for _, c := range candidatos {
				if c == url {
					encontrado = true
					break
				}
			}
			if !encontrado {
				candidatos = append(candidatos, url)
			}
		}
	}

	return candidatos
}

// iniciarHLS inicia streaming RTSP→HLS para una cámara
func iniciarHLS(indiceOID interface{}, urlRTSP interface{}, config map[string]interface{}, velocidadKbps *float64) (bool, string) {
	camID := resolverCamID(indiceOID)
	if camID == "" {
		return false, "Identificador de cámara inválido"
	}

	// Recopilar candidatos
	var candidatos []string
	switch v := urlRTSP.(type) {
	case []string:
		candidatos = v
	case string:
		if v != "" {
			candidatos = []string{v}
		}
	}

	if len(candidatos) == 0 {
		return false, "URL RTSP vacía"
	}

	ultimoError := "URL RTSP no válida"
	for idx, url := range candidatos {
		ok, msg := iniciarHLSSingle(camID, url, config, velocidadKbps)
		if ok {
			if idx > 0 {
				msg = fmt.Sprintf("%s (usando candidato %d)", msg, idx+1)
			}
			return true, msg
		}
		ultimoError = msg
	}
	return false, ultimoError
}

// iniciarHLSSingle lanza ffmpeg para una URL específica
func iniciarHLSSingle(camID, urlRTSP string, config map[string]interface{}, velocidadKbps *float64) (bool, string) {
	// Verificar ffmpeg
	ffmpegPath, err := exec.LookPath("ffmpeg")
	if err != nil {
		return false, "ffmpeg no encontrado en el sistema"
	}

	hlsAsegurarCarpetas()

	// Detener si ya hay un proceso activo
	detenerHLS(camID)

	carpeta := filepath.Join(hlsOutputDir, camID)
	limpiarSalidaHLS(camID)

	playlistName, ok := hlsPlaylistName[camID]
	if !ok {
		playlistName = camID + ".m3u8"
	}
	playlistPath := filepath.Join(carpeta, playlistName)
	segmentPattern := filepath.Join(carpeta, camID+"_%03d.ts")

	// Obtener resolución
	resolucion := obtenerResolucion(config, camID, velocidadKbps)
	resConfig, ok := resolucionConfig[resolucion]
	if !ok {
		resConfig = resolucionConfig["480p"]
	}

	log.Printf("HLS %s: usando %s (%s, %s)", camID, resolucion, resConfig.Scale, resConfig.Bitrate)

	cmd := exec.Command(ffmpegPath,
		"-nostdin",
		"-loglevel", "error",
		"-rtsp_transport", "tcp",
		"-i", urlRTSP,
		"-an",
		"-vf", "scale="+resConfig.Scale,
		"-c:v", "libx264",
		"-preset", "ultrafast",
		"-b:v", resConfig.Bitrate,
		"-maxrate", resConfig.Maxrate,
		"-bufsize", "1000k",
		"-f", "hls",
		"-hls_time", "0.5",
		"-hls_list_size", "3",
		"-hls_flags", "delete_segments+program_date_time",
		"-hls_segment_filename", segmentPattern,
		playlistPath,
	)

	if err := cmd.Start(); err != nil {
		return false, "Error iniciando ffmpeg: " + err.Error()
	}

	procesosHLSMu.Lock()
	procesosHLS[camID] = cmd
	procesosHLSMu.Unlock()

	log.Printf("Iniciando HLS %s desde %s (PID %d)", camID, urlRTSP, cmd.Process.Pid)

	// Esperar hasta 3 segundos para verificar que el proceso no terminó inmediatamente
	for i := 0; i < 6; i++ {
		time.Sleep(500 * time.Millisecond)

		// Verificar si el proceso terminó
		if cmd.ProcessState != nil {
			procesosHLSMu.Lock()
			delete(procesosHLS, camID)
			procesosHLSMu.Unlock()
			return false, "No se pudo abrir el RTSP (proceso terminó)"
		}

		// Verificar si el archivo playlist existe
		if info, err := os.Stat(playlistPath); err == nil && info.Size() > 0 {
			return true, "HLS iniciado"
		}
	}

	return true, "HLS iniciado (validación pendiente)"
}

// detenerHLS detiene el ffmpeg de una cámara
func detenerHLS(indiceOID interface{}) bool {
	camID := resolverCamID(indiceOID)
	if camID == "" {
		return false
	}

	procesosHLSMu.Lock()
	cmd, ok := procesosHLS[camID]
	if !ok {
		procesosHLSMu.Unlock()
		return true
	}
	delete(procesosHLS, camID)
	procesosHLSMu.Unlock()

	if cmd.Process != nil {
		cmd.Process.Signal(os.Interrupt)
		done := make(chan error, 1)
		go func() {
			done <- cmd.Wait()
		}()

		select {
		case <-done:
		case <-time.After(5 * time.Second):
			cmd.Process.Kill()
		}
	}

	log.Printf("HLS detenido para %s", camID)
	return true
}

// hlsDetenerTodos detiene todos los procesos ffmpeg
func hlsDetenerTodos() {
	procesosHLSMu.Lock()
	camIDs := make([]string, 0, len(procesosHLS))
	for id := range procesosHLS {
		camIDs = append(camIDs, id)
	}
	procesosHLSMu.Unlock()

	for _, id := range camIDs {
		detenerHLS(id)
	}
}

// ==================== UTILIDADES ====================

func resolverCamID(indiceOID interface{}) string {
	switch v := indiceOID.(type) {
	case string:
		if v == "cam1" || v == "cam2" {
			return v
		}
	case int:
		if v == 1 {
			return "cam1"
		}
		if v == 2 {
			return "cam2"
		}
	case float64:
		if v == 1 {
			return "cam1"
		}
		if v == 2 {
			return "cam2"
		}
	}
	return ""
}

func limpiarSalidaHLS(camID string) {
	carpeta := filepath.Join(hlsOutputDir, camID)
	entries, err := os.ReadDir(carpeta)
	if err != nil {
		return
	}
	for _, entry := range entries {
		nombre := entry.Name()
		if strings.HasSuffix(nombre, ".ts") || strings.HasSuffix(nombre, ".m3u8") {
			os.Remove(filepath.Join(carpeta, nombre))
		}
	}
}

// ==================== DESCUBRIMIENTO ONVIF ====================

// descubrirONVIF realiza una búsqueda WS-Discovery de cámaras ONVIF en la red
func descubrirONVIF(timeoutSec int) []map[string]interface{} {
	if timeoutSec <= 0 {
		timeoutSec = 3
	}

	messageID := fmt.Sprintf("uuid:%d", time.Now().UnixNano())
	probe := fmt.Sprintf(`<?xml version='1.0' encoding='UTF-8'?>
<e:Envelope xmlns:e='http://www.w3.org/2003/05/soap-envelope'
            xmlns:w='http://schemas.xmlsoap.org/ws/2004/08/addressing'
            xmlns:d='http://schemas.xmlsoap.org/ws/2005/04/discovery'
            xmlns:dn='http://www.onvif.org/ver10/network/wsdl'>
  <e:Header>
    <w:MessageID>%s</w:MessageID>
    <w:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</w:To>
    <w:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Probe</w:Action>
  </e:Header>
  <e:Body>
    <d:Probe><d:Types>dn:NetworkVideoTransmitter</d:Types></d:Probe>
  </e:Body>
</e:Envelope>`, messageID)

	addr, err := net.ResolveUDPAddr("udp4", "239.255.255.250:3702")
	if err != nil {
		log.Printf("Error resolviendo multicast: %v", err)
		return []map[string]interface{}{}
	}

	conn, err := net.ListenUDP("udp4", nil)
	if err != nil {
		log.Printf("Error abriendo socket UDP: %v", err)
		return []map[string]interface{}{}
	}
	defer conn.Close()

	conn.SetDeadline(time.Now().Add(time.Duration(timeoutSec) * time.Second))

	_, err = conn.WriteToUDP([]byte(probe), addr)
	if err != nil {
		log.Printf("Error enviando probe ONVIF: %v", err)
		return []map[string]interface{}{}
	}

	dispositivos := map[string]map[string]interface{}{}
	buf := make([]byte, 8192)

	for {
		n, remoteAddr, err := conn.ReadFromUDP(buf)
		if err != nil {
			break
		}

		host := remoteAddr.IP.String()
		txt := string(buf[:n])

		// Buscar XAddrs
		xaddrs := ""
		if idx := strings.Index(txt, "XAddrs"); idx >= 0 {
			// Encontrar inicio del contenido (después del >)
			start := strings.Index(txt[idx:], ">")
			if start >= 0 {
				end := strings.Index(txt[idx+start:], "<")
				if end >= 0 {
					xaddrs = strings.TrimSpace(txt[idx+start+1 : idx+start+end])
				}
			}
		}

		// Inferir puerto desde XAddrs
		puerto := 8899
		if colonIdx := strings.Index(xaddrs, "://"); colonIdx >= 0 {
			resto := xaddrs[colonIdx+3:]
			if portIdx := strings.Index(resto, ":"); portIdx >= 0 {
				portEnd := strings.IndexAny(resto[portIdx+1:], "/ ")
				if portEnd < 0 {
					portEnd = len(resto[portIdx+1:])
				}
				if p, err := strconv.Atoi(resto[portIdx+1 : portIdx+1+portEnd]); err == nil {
					puerto = p
				}
			}
		}

		dispositivos[host] = map[string]interface{}{
			"host":   host,
			"puerto": puerto,
			"xaddrs": xaddrs,
		}
	}

	resultado := make([]map[string]interface{}, 0, len(dispositivos))
	for _, d := range dispositivos {
		resultado = append(resultado, d)
	}
	return resultado
}
